### ISSUE TITLE
TC errors reported that are wrong

### ISSUE THREAD

**Issue by joey-coleman** [symphonytool/symphony#107]
> The following bug was [originally reported](https://sourceforge.net/p/compassresearch/bugs/107/) on Sourceforge by [pglvdm](https://sourceforge.net/u/pglvdm/profile/), 2013-10-08 14:33:56.585000:

In the C:\compasssvn\Common\CaseStudies\GM_01\DeviceBattery.cml file the TC reports a number of errors. The first 4 claims that nat is not a subset of real. This error message is definitely wrong. In addition all types derived here should be reals. The 5th error message here seem to think that the type to be communicated over is of type Time and not as it is declares a pair where Time is the first element.

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/107/#e41b) by [joeycoleman](https://sourceforge.net/u/joeycoleman/profile/), 2013-10-09 08:21:28.400000:
- **assigned_to**: Claus B. Nielsen

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/107/#0bf0) by [clausbn](https://sourceforge.net/u/clausbn/profile/), 2013-10-11 15:44:12.584000:
- **status**: open --> accepted

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/107/#480f) by [clausbn](https://sourceforge.net/u/clausbn/profile/), 2013-10-11 15:44:12.948000:

Reals are now accepted in relation to numeric types. 

[992ab9f13d19013ce7b64921d20057ddcc6fb274]

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/107/#7bad) by [clausbn](https://sourceforge.net/u/clausbn/profile/), 2013-10-11 16:56:48.181000:

There is some confusion as to how product types should be used on channels. 

If we have a channel chan : (nat \* nat)

The vdm oriented people use them as chan!mk_(x,y) (GM example), while the CSP oriented people use them as chan?x!y, (Library example)

Currently it seems the TC expected the latter and therefore assumes the first channel parameter to be the first type in the product.

Should both be allowed? If so, it becomes difficult for the TC to know if it should expect a product type or it should expect each parameter to match the unwrapped channel product type.

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/107/#4b0c) by [joeycoleman](https://sourceforge.net/u/joeycoleman/profile/), 2013-10-15 08:42:54.610000:

I remember discussing this with Anders and Rasmus last year; we couldn't figure out what the difference should be between two channels defined as

```
channels
  a : int * int
  b : (int * int)
```

Nor, for that matter, were we able to put our collective finger on the precise difference between

```
b!1!2 -> P
```

and
    a!mk_(1,2) -> P

beyond the obvious change in channel names.

However, there is a suspicious echo between this and the ABracketType that the function/operation types needed.

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/107/#09b0) by [lausdahl](https://sourceforge.net/u/lausdahl/profile/), 2013-11-04 14:12:49.709000:
- **Module**:  --> type-checker

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/107/#e9c3) by [joeycoleman](https://sourceforge.net/u/joeycoleman/profile/), 2013-11-05 12:45:28.466000:

Suggested resolution: we should treat channel types like function/operation types.

so, for

```
types
  T = int * int

channels
  a : int * int
  b : (int * int)
  c : T
```

the types of channels `b` and `c` are the same, but the type of `a` differs.

Synchronisation with channel `a` is of the form: `a!1!2 -> P`,
whereas for `b` and `c` it looks like: `b!mk_(1,2) -> P`.

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/107/#9b58) by [lausdahl](https://sourceforge.net/u/lausdahl/profile/), 2013-11-06 23:31:00.892000:

Fixed in [9d11c519e643c371fb966284df5e4a087d880ce3] a channel type has been implemented to be able to hold the list of types. The communication action check is changed to reject applications when arguments mismatch

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/107/#fea3) by [lausdahl](https://sourceforge.net/u/lausdahl/profile/), 2013-11-06 23:31:15.107000:
- **status**: accepted --> closed
